artiq/coredevice/mirny.py

```diff
@@ -140,3 +151,232 @@ def write_ext(self, addr, length, data):
         if length < 32:
             data <<= 32 - length
         self.bus.write(data)
+
+
+class Almazny:
+    """
+    Almazny (High frequency mezzanine board for Mirny)
+
+    :param host_mirny - Mirny device Almazny is connected to
+    """
+
+    def __init__(self, dmgr, host_mirny):
+        self.mirny = dmgr.get(host_mirny)
+        self.att_mu = [0x3f] * 4
+        self.rf_switches = [0] * 4
+
+    @kernel
+    def init(self):
+        self._send_mezz_data([
+            (ALMAZNY_REG_CLEAR, 1),
+            # rest is 0
+            ])
+
+    @kernel
+    def _reverse_att_mu(self, att_mu):
+        res = 0
+        for i in range(6):
+            res |= (att_mu >> i & 0x01) << (5-i)
+        return res
+
+    @kernel
+    def att_to_mu(self, att):
+        """
+        Convert an attenuator setting in dB to machine units.
+
+        :param att: attenuator setting in dB [0-31.5] in 0.5 steps
+        :return: attenuator setting in machine units
+        """
+        if att > 31.5 or att < 0 or (att-int(att) != 0.0 and att-int(att) != 0.5):
+            raise ValueError("Invalid Almazny attenuator settings!")
+        # it must be reversed for 
+        return self._reverse_att_mu(int(att * 2))
+
+    @kernel
+    def mu_to_att(self, att_mu):
+        """
+        Convert a digital attenuator setting to dB.
+
+        :param att_mu: attenuator setting in machine units
+        :return: attenuator setting in dB
+        """
+        return self._reverse_att_mu(att_mu) / 2
+
+    @kernel
+    def set_att_mu(self, channel, att_mu):
+        """
+        Sets attenuators on chosen shift register (channel).
+        :param channel - index of the register [1-4]
+        :param att_mu - attenuation setting in machine units [0-63]
+        """
+        if not 4 >= channel >= 1:
+            raise ValueError("channel must be 1, 2, 3 or 4")
+        if not 63 >= att_mu >= 0:
+            raise ValueError("Invalid Almazny attenuator setting")
+
+        ch = channel - 1
+        self.att_mu[ch] = att_mu
+        self._update_register(ch)
+
+    @kernel
+    def set_att_mu_all(self, att_mu):
```

Comment:
Other boards don't have such a method and I'm not sure if there's a use case for it. Very disparate equipment, requiring different power levels, could be connected to each Almazny channel.

Reply:
Urukul does have a method like that, that's what I based this on.

Reply:
Look closer - the Urukul method does something else: it lets you set different values for each attenuator. And please use exactly the same method names.